### ISSUE TITLE
aws 23.0.0: broken build on Fedora

### ISSUE THREAD

**Issue by mosteo** [alire-project/alire#1394]
See https://github.com/alire-project/alire-index/actions/runs/4720095866/jobs/8372670729?pr=795

**Comment by stcarrez** [comment]
Some notes when trying to spot the issue.... It is not related to Fedora and it occurs on other systems.  I've created a blank project that only `with aws` and at some point I got the following error:

```
   [mkdir]        object directory for project Valhalla
   [mkdir]        exec directory for project Valhalla
src.gpr:39:37: "../.build/x86_64-pc-linux-gnu/setup/src" is not a valid directory
gprbuild: "/src/vacs/valhalla/valhalla.gpr" processing failed
error: Command ["gprbuild", "-s", "-j0", "-p", "-P", "/src/vacs/valhalla/valhalla.gpr"] exited with code 4
error: Compilation failed.
```

Looking at the `cache/dependencies/aws_23.0.0_6c406859/.build` directory, it contained a
`x86_64-linux-gnu` and another `x86_64-pc-linux-gnu` directory. The `src` directory was generated in `x86_64-linux-gnu`.  The `makefile.setup` was created with the line:

```
TARGET=x86_64-linux-gnu
```

but the AWS toml configuration is defined with:

```
TARGET = "x86_64-pc-linux-gnu"
```

I think the root issue is that when we run the AWS `post-fetch` setup command, we don't give the `TARGET` configuration. It means the configuration is made by using AWS default which is:

```
TARGET = $(shell gcc -dumpmachine)
```

When Alire uses its own compiler, you get:

```
~/.config/alire/cache/dependencies/gnat_native_12.1.2_66b989b5/bin/gcc -dumpmachine
x86_64-pc-linux-gnu
```

But if you use some system default, you get:

```
/usr/bin/gcc -dumpmachine
x86_64-linux-gnu
```

which means that depending on your PATH and which compiler you use and/or is installed on the system by default, we can get different results.

@Fabien-Chouteau , to solve this issue, we could change the `post-fetch` action to:

```
command = ["make", "setup", "ZLIB=false", "DEMOS=false",
           "LAL=false", "XMLADA=false", "SOCKET=openssl", "prefix=install_dir",
           "PRJ_BUILD=Release", TARGET="x86_64-pc-linux-gnu"]
```

which also means there must be specific `post-fetch` actions for each target.

This explains the failure of [Ada Utility Library AWS 2.6.0 crate](https://github.com/alire-project/alire-index/pull/888) for Debian and Fedora (it's ok for Arch and Ubuntu latest).

**Comment by stcarrez** [comment]
@Fabien-Chouteau and @mosteo, I think the root issue comes from the script https://github.com/alire-project/alire-index-checks/blob/stable-1.2.1/scripts/gh-build-crate.sh.

What happens is that the `alr` dependencies are run first with a default compiler in the PATH (`/usr/bin/gcc` probably).

Then, we seem to select a different compiler with `alr toolchain --select` and a different compiler is used: the `TARGET` is changed but the `setup` scripts have been executed.

**Comment by mosteo** [comment]
Hi @stcarrez, I'm trying to understand the issue but falling short of it. I think I get the implications you mention about running with default environment and explicit toolchain the actions vs actual build. I'm not sure though about what can be done in the `gh-build-crate.sh`. Do you see a simple change that could fix the issue?

In your previous message you said that a simple `alr aws` would trigger the problem, which seems to contradict that the root cause be in `gh-build-crate.sh`?

(Finally, with `alr 2.0` being hopefully not far away, bringing improvements in the build process, it might be fixed there. For example, the actions should all run with the same  environment as the actual build. So it might be worth a try to check if you can reproduce that simple way of triggering the problem with an alr nightly build.)

**Comment by stcarrez** [comment]
The issue is triggered if the `aws` setup is run with a first gnat compiler in the path but another gnat compiler is used when building. The `aws` setup will choose a first `TARGET` which will dictate the gnat compiler to use by `aws` GNAT projects but it is not in the path anymore due to some setup switch.

I've looked at the `gh-build-crate.sh` script together with the failed build outputs but it is difficult to say what triggered the gnat compiler change.

I wanted to run `gh-build-crate.sh` manually to check but didn't had time to investigate which arguments are necessary and how to launch it.  If you tell me how I can launch it, I could investigate further on my side.

I suspect some different environment between `arch` where it builds correctly and others that have the issue.

**Comment by mosteo** [comment]
I tried to reproduce with AWS 24 but it has been fixed there, as it builds with either Alire's or Fedora's compiler.

With AWS 23 and `alr 2.0` and the default Alire compiler there's no problem either. When trying to use the Fedora compiler the issue triggers again, and I don't see how there can be two compilers involved in that case, as only the Fedora one is visible to Alire. This is in a clean build, not trying to mess things on purpose, so I don't know...

Given the fix with AWS 24 I'm inclined to not lose more sleep over this, so I'm closing this issue.
